File: study/ts/script/scene/view/historyhero/HistoryHeroAttrLayer.ts
import { HistoryHeroConst } from "../../../const/HistoryHeroConst";
import CommonCustomListView from "../../../ui/component/CommonCustomListView";
import CommonDetailWindow from "../../../ui/component/CommonDetailWindow";
import { HistoryHeroDataHelper } from "../../../utils/data/HistoryHeroDataHelper";
import { TypeConvertHelper } from "../../../utils/TypeConvertHelper";
import HistoryHeroDetailAttrModule from "./HistoryHeroDetailAttrModule";
import HistoryHeroDetailAwakenModule from "./HistoryHeroDetailAwakenModule";
import HistoryHeroDetailDescModule from "./HistoryHeroDetailDescModule";
import HistoryHeroDetailSkillModule from "./HistoryHeroDetailSkillModule";
import HistoryHeroDetailWeaponModule from "./HistoryHeroDetailWeaponModule";

const { ccclass, property } = cc._decorator;


@ccclass
export default class HistoryHeroAttrLayer extends cc.Component {
    @property({
        type: CommonDetailWindow,
        visible: true
    })
    _commonDetailWindow: CommonDetailWindow = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _name2: cc.Label = null;

    @property({
        type: CommonCustomListView,
        visible: true
    })
    _listView: CommonCustomListView = null;

    @property(HistoryHeroDetailAttrModule)
    attrModule: HistoryHeroDetailAttrModule = null;
    @property(HistoryHeroDetailSkillModule)
    skillModule: HistoryHeroDetailSkillModule = null;

    @property(HistoryHeroDetailWeaponModule)
    weaponModule: HistoryHeroDetailWeaponModule = null;

    @property(HistoryHeroDetailAwakenModule)
    awakenModule: HistoryHeroDetailAwakenModule = null;

    @property(HistoryHeroDetailDescModule)
    descModule: HistoryHeroDetailDescModule = null;
    _weaponData: any;


    onEnter() {
    }
    onExit() {
    }
    updateUI(data) {
        if (data == null) {
            return;
        }
        this._listView.removeAllChildren();
        this._updateAttrModule(data.getSystem_id(), data.getBreak_through());
        this._updateSkillDesc(data.getSystem_id(), data.getBreak_through());
        this._updateWeapon(data.getSystem_id());
        this._updateAwaken(data.getSystem_id());
        this._updateDesc(data.getSystem_id());
    }
    updateUIForNoraml(systemId) {
        if (systemId == null) {
            return;
        }
        this._listView.removeAllChildren();
        this._updateAttrModule(systemId, 1);
        this._updateSkillDesc(systemId, 1);
        this._updateWeapon(systemId);
        this._updateAwaken(systemId);
        this._updateDesc(systemId);
    }
    _updateAttrModule(baseId, breakthrough) {
        var stepCfg = HistoryHeroDataHelper.getHistoryHeroStepByHeroId(baseId, breakthrough);
        if (stepCfg == null) {
            return;
        }
        this.attrModule.updateUI(stepCfg);
        this._listView.pushBackCustomItem(this.attrModule.node);
    }
    _updateSkillDesc(baseId, breakthrough) {
        var skilist = HistoryHeroDataHelper.getHistoricalSkills(baseId);
        if (skilist.length > 0) {
            this.skillModule.ctor(skilist, breakthrough);
            this._listView.pushBackCustomItem(this.skillModule.node);
        }
    }
    _updateWeapon(baseId) {
        this.weaponModule.ctor(baseId);
        this._listView.pushBackCustomItem(this.weaponModule.node);
    }
    _updateAwaken(baseId) {
        var heroInfoCfg = HistoryHeroDataHelper.getHistoryHeroInfo(baseId);
        if (heroInfoCfg.color == HistoryHeroConst.QUALITY_ORANGE) {
            this.awakenModule.node.active = true;
            var costList = HistoryHeroDataHelper.getHistoricalAwakenCostList(baseId);
            this.awakenModule.ctor(costList);
            this._listView.pushBackCustomItem(this.awakenModule.node);
        }
    }
    _updateDesc(baseId) {
        var historyHeroInfo = HistoryHeroDataHelper.getHistoryHeroInfo(baseId);
        this.descModule.ctor(historyHeroInfo);
        this._listView.pushBackCustomItem(this.descModule.node);
    }

    setName(heroBaseId) {
        var heroParam = TypeConvertHelper.convert(TypeConvertHelper.TYPE_HISTORY_HERO, heroBaseId);
        this._name2.string = (heroParam.name);
        this._name2.node.color = (heroParam.icon_color);
    }

}

File: study/lua/src/app/scene/view/historyhero/HistoryHeroAttrLayer.lua
-- Author: conley
-- Date:2018-11-23 17:08:00
-- @Rebuild by Panhoa
-- Describle：

local ViewBase = require("app.ui.ViewBase")
local HistoryHeroAttrLayer = class("HistoryHeroAttrLayer", ViewBase)
local HistoryHeroDataHelper = require("app.utils.data.HistoryHeroDataHelper")
local HistoryHeroDetailAttrModule = require("app.scene.view.historyhero.HistoryHeroDetailAttrModule")
local HistoryHeroDetailSkillModule = require("app.scene.view.historyhero.HistoryHeroDetailSkillModule")
local HistoryHeroDetailDescModule = require("app.scene.view.historyhero.HistoryHeroDetailDescModule")
local HistoryHeroDetailWeaponModule = require("app.scene.view.historyhero.HistoryHeroDetailWeaponModule")
local HistoryHeroDetailAwakenModule = require("app.scene.view.historyhero.HistoryHeroDetailAwakenModule")


function HistoryHeroAttrLayer:ctor()
	self._propertyNode = nil

	local resource = {
		file = Path.getCSB("HistoryHeroAttrLayer", "historyhero"),

	}
	HistoryHeroAttrLayer.super.ctor(self, resource)
end

function HistoryHeroAttrLayer:onCreate()
	self._listViewContentSize = self._listView:getContentSize()
end

function HistoryHeroAttrLayer:onEnter()
end

function HistoryHeroAttrLayer:onExit()
end

function HistoryHeroAttrLayer:setNodeVisible(bVisible)
	self._propertyNode:setVisible(bVisible)
end

-- @Role 	适配详情界面
function HistoryHeroAttrLayer:updateScrollHeight()
	self._listView:setContentSize(self._listViewContentSize.width, self._listViewContentSize.height)
end

-- @Role
function HistoryHeroAttrLayer:updateUI(data)
	if data == nil then
		return
	end

	self._listView:removeAllChildren()
	self:_updateAttrModule(data:getSystem_id(), data:getBreak_through())
	self:_updateSkillDesc(data:getSystem_id(), data:getBreak_through())
	self:_updateWeapon(data:getSystem_id())
	self:_updateAwaken(data:getSystem_id())
	self:_updateDesc(data:getSystem_id())
end

-- @Role 
function HistoryHeroAttrLayer:updateUIForNoraml(systemId)
	if systemId == nil then
		return
	end

	self._listView:removeAllChildren()
	self:_updateAttrModule(systemId, 1)
	self:_updateSkillDesc(systemId, 1)
	self:_updateWeapon(systemId)
	self:_updateAwaken(systemId)
	self:_updateDesc(systemId)
end

-- @Role
function HistoryHeroAttrLayer:_updateAttrModule(baseId, breakthrough)
	local stepCfg = HistoryHeroDataHelper.getHistoryHeroStepByHeroId(baseId, breakthrough)
	if stepCfg == nil then
		return
	end
	local attrModule = HistoryHeroDetailAttrModule.new()
	self._listView:pushBackCustomItem(attrModule)
	attrModule:updateUI(stepCfg)
end

-- @Role
function HistoryHeroAttrLayer:_updateSkillDesc(baseId, breakthrough)
	local skilist = HistoryHeroDataHelper.getHistoricalSkills(baseId)
	if #skilist > 0 then
		local skillModule = HistoryHeroDetailSkillModule.new(skilist, breakthrough)
		self._listView:pushBackCustomItem(skillModule)
	end
end

-- @Role
function HistoryHeroAttrLayer:_updateWeapon(baseId)
	local weaponModule = HistoryHeroDetailWeaponModule.new(baseId)
	self._listView:pushBackCustomItem(weaponModule)
end

-- @Role
function HistoryHeroAttrLayer:_updateAwaken(baseId)
	local HistoryHeroConst = require("app.const.HistoryHeroConst")
	local heroInfoCfg = HistoryHeroDataHelper.getHistoryHeroInfo(baseId)
	if heroInfoCfg.color == HistoryHeroConst.QUALITY_ORANGE then
		local costList = HistoryHeroDataHelper.getHistoricalAwakenCostList(baseId)
		local awakenModule = HistoryHeroDetailAwakenModule.new(costList)
		self._listView:pushBackCustomItem(awakenModule)
	end
end

-- @Role
function HistoryHeroAttrLayer:_updateDesc(baseId)
	local historyHeroInfo = HistoryHeroDataHelper.getHistoryHeroInfo(baseId)
	local descModule = HistoryHeroDetailDescModule.new(historyHeroInfo)
	self._listView:pushBackCustomItem(descModule)
end

return HistoryHeroAttrLayer